File: Week_06/152/maximum_product_subarray.py
from typing import List


class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        ans = max_product = min_product = nums[0]
        for num in nums[1:]:
            m, n = max_product, min_product
            max_product = max(m * num, n * num, num)
            min_product = min(m * num, n * num, num)
            ans = max(ans, max_product)
        return ans

File: Week_06/152/maximum_product_subarray.go
package main

func maxProduct(nums []int) int {
	ans, maxF, minF := nums[0], nums[0], nums[0]
	for i := 1; i < len(nums); i++ {
		m, n := maxF, minF
		maxF = max(m*nums[i], max(n*nums[i], nums[i]))
		minF = min(m*nums[i], min(n*nums[i], nums[i]))
		ans = max(maxF, ans)
	}
	return ans
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}
